xcpmd/src/battery.c

```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
         return NORMAL;
 }
 
+
+//Returns whether a battery is charging, discharging, full, etc.
+int get_battery_charge_state(unsigned int battery_index) {
+
+    unsigned int percent, i;
+
+    if (last_status[battery_index].state & 0x1)
+        return BATT_DISCHARGING;
+    else if (last_status[battery_index].state & 0x2)
+        return BATT_CHARGING;
+    else {
+        /* We're not charging nor discharging... */
+        percent = get_battery_percentage(battery_index);
+        /* Are we full or empty? */
+        if (percent > 90)
+            return BATT_FULL;
+        else if (percent < 10)
+             return BATT_EMPTY;
+        else {
+            /* Is anybody else (dis)charging? */
+            for (i = 0; i < num_battery_structs_allocd; ++i) {
+                if (i != battery_index &&
+                    last_status[i].present == YES &&
+                    (last_status[i].state & 0x1 || last_status[i].state & 0x2)) {
+                    break;
+                }
+            }
+            if (i < num_battery_structs_allocd) {
+                /* Yes! */
+                /* If the other battery is charging, we're pending charge */
+                if (last_status[i].state & 0x2)
+                    return BATT_PENDING_CHARGE;
+                /* If the other battery is discharging, we're pending discharge */
+                else if (last_status[i].state & 0x1)
+                    return BATT_PENDING_DISCHARGE;
+            } else
+                /* We tried everything, the state is unknown... */
+                return BATT_STATE_UNKNOWN;
+        }
+    }
+
+    return BATT_STATE_UNKNOWN;
+}
+
+
+//Returns whether the system as a whole is charging, discharging, full, or empty.
+int get_system_charge_state(void) {
+
+    int percentage = get_overall_battery_percentage();
+    long charge_rate = get_total_charge_rate();
+
+    if (get_ac_adapter_status() == ON_AC) {
+        if (percentage > 98 || charge_rate == 0) {
```

Comment:
The idea was knowing the charge rate gives us a much greater degree of certainty that the batteries really are full. The percentage was thrown in as a fallback for systems that still report a trickle of battery current even when things are full. That aside, you're right in that there should be consistency--I'll correct this.